### ISSUE TITLE
It's very often when Web3modal isn't being opened after an openModal call

### ISSUE THREAD

**Issue by alexeygorbachevskiy** [WalletConnect/web3modal#776]
Breef code example:
```
try {
const  walletConnectClient = await WalletConnectClient.init({
          projectId: WALLET_CONNECT_PROJECT_ID,
});

const web3modal = new Web3Modal({
          projectId: WALLET_CONNECT_PROJECT_ID,
          standaloneChains: [WALLET_CONNECT_CHAINS.stellarPubnet],
});

const { uri, approval } = await walletConnectClient.connect({
          requiredNamespaces: {
            [WALLET_CONNECT_CHAINS.stellarPubnet.split(":")[0]]: {
              chains: [WALLET_CONNECT_CHAINS.stellarPubnet],
              methods: [WALLET_CONNECT_METHODS.signTx],
              events: [],
            },
          },
});

web3modal.openModal({
      uri,
      standaloneChains: [WALLET_CONNECT_CHAINS.stellarPubnet],
});

walletConnectSession = await approval();
web3modal.closeModal();
} 
catch(e) {
// do smth
}

```

Actual Result: 
The modal is not opened.
![image](https://user-images.githubusercontent.com/37103509/208078414-6235824d-d28b-4420-b9f6-4b322e0b3a3b.png)

**Comment by 0xAsimetriq** [comment]
To clarify, you are using `@web3modal/standalone`, latest release?

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq 
I am using 2.0.0-beta.9 version of @web3modal/standalone. The flow itself is very simple: Just click to signIn button which calls the code above. And it's very often when modal is not being opened, but sometimes this is being opened - I don't see the reasons of such behavior nor log errors nor so on.

**Comment by 0xAsimetriq** [comment]
@alexeygorbachevskiy I think I know what's up, there was likely some condition where sign client was not initialised. I created `html-standalone` example, that you can reference here
https://github.com/WalletConnect/web3modal/tree/beta.10/examples/html-standalone

Note how it initialises `signClient` and also checks for it to be defined before opening the modal.

To provide nicer ux, you can also disable / show some loading state while clients are initialising.

Let me know if following that example solves your issue.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq  But, if client is not been initialized, then wc has issue with initialization anyway..

**Comment by 0xAsimetriq** [comment]
I don't entirely understand what you mean by that?
TL;DR

You need to initialise `SignClient` first. Once it is ready you can call connect method on it to get `uri` for pairing. Modal needs this uri to work properly, once you have it you can open the modal.

If you try to open the modal before you have it, there will likely be some errors / it won't open as there is no data to display in qrcode or add to deeplinks.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq 
See code above with attention pls, I open modal after initialization
![image](https://user-images.githubusercontent.com/37103509/208108819-a3223246-b834-4f60-baac-dd765125ab34.png)

**Comment by 0xAsimetriq** [comment]
I'll need to see full repo, to verify how snippet above works. Meanwhile, here are some ideas

1. Is your project id 100% defined?
2. Where is `pairing?.topic` coming from, is it correct?
3. I'd add a console log before `web3modal.openModal` call, to check if code actually gets to that stage and what uri is
4. If you are using existing pairing and it exists, uri won't be there, you need to check for that and not even open the modal in that case

**Comment by alexeygorbachevskiy** [comment]
1. Yep
2. U can not consider `pairing?.topic` since it equals `undefined` in this case
3. The web3modal.openModal  is called but nothing happened
4. I don't use existing pairings

And pls, again, check with attention my initial post - I sent this screen cause it will appear only if web3modal.openModal was called:

![image](https://user-images.githubusercontent.com/37103509/208130287-9040c294-d35f-43f8-ad18-dae77e0e8301.png)

**Comment by 0xAsimetriq** [comment]
`w3m-container` is always present on a page, content within it get's appended once modal is open. I can't reproduce this issue at the moment (see vid below), so will need to spin up a local reproduction to debug it further. Is your repo open by any chance? Or could you add me to it if it is private?

(video is using that `html-standalone` example I mentioned above, I'd start there and see what differences you have with it)

https://user-images.githubusercontent.com/3154053/208131623-e4531ade-f700-4344-843e-4bf544f3b19a.mov

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq 
It seems to be an initialization (`await WalletConnectClient.init`) problem that just waits a long time - very long time, just hanging

**Comment by 0xAsimetriq** [comment]
Hmm, it should take a while, but no more than a few seconds. Does this page take long for you to load as well https://react-wallet.walletconnect.com ?

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq I have rechecked and if I write console.log straight before `web3modal.openModal` call, then this console.log called, but modal is not opened.  As the code itself go to this point, so it seems the issue in `web3modal.openModal`  itself.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq How are you?

**Comment by 0xAsimetriq** [comment]
Still unable to re-produce this, so am thinking there are some details missing, but it is hard to pinpoint. If you could link full reproduction repo, I can deep dive and have a look. Could be some sneaky networking relate race condition.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq I can't share cause I have private repo. But it seems the issue occurres when we make login->logout->login via web3 modal - try cases smth like this pls.

**Comment by 0xAsimetriq** [comment]
Are you re-using existing sign client when doing that or initialising it every time new connection is made? That might cause some issues.

1. You should initialise sign client only once per lifecycle app
2. Same goes for web3modal
3. When re-connecting you only need to call `.connect` method to obtain new pairing uri (or you can re-use existing)

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq 
Yep, I reuse existing sign client. It seems the issue occurres when we made some logins-logouts with existing sign client and after reload page try to open modal again.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq I have created private repo with minimum reproducable code. Accept invite pls and try to investigate. Don't forget to provide _**REACT_APP_WALLET_CONNECT_PROJECT_ID**_ env var in `.env` file if u wnt to test it locally.
I have noticed that issue reproduces more when app is deployed rathen than on local env.
Netlify link: https://wallet-connect-web3-modal.netlify.app/

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq Have you seen my repo?

**Comment by 0xAsimetriq** [comment]
@alexeygorbachevskiy I think it has something to do with different "browserslist" list options in `package.json`

locally everything works fine for me, the only issue I see is that when there are many pairings it takes a while for everything to load (we are working on this one), but modal opens up every time.

If you apply production browserlist to development, it breaks (remember to restart compiler), hence why I think you are seeing issues in deployed apps.

I believe common broserlist config these days is `defaults and supports es6-module` you could test that.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq Yep, I've reprodced it. This is exactly fails by the reason of browserlist config. First of all I suggest to return back label `issue` for this thread. It seems to me that it's an issue when we can't use common prod browserlist config: 
```
"production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ]
```
It is really common prod config. For ex. `cra` also use it as a default for prod.

**Comment by 0xAsimetriq** [comment]
We are targeting es2020 spec, which includes es6-modules that are all widely adopted these days. Since this issue is solved by updating cra's config to a more modern one and everything works out of the box in other tools like vite / latest babel and nextjs, I think this issue can be closed, as it is not web3modal specific, rather we need to include this info in the docs for future cra developers to be aware of.

Glad we found the root cause 👍

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq Hm, I've deployed my repo (mentioned above) with `defaults and supports es6-module` browserlist setings and issue is still being reproduced. Can you check it out?

**Comment by 0xAsimetriq** [comment]
Works locally for me (both dev and prod builds). But deployed example above indeed fails, few ideas

1. Check for any caches used during the build
2. It seems that it always works if you wait for 5s or so before clicking the button, this leads me to think that there might be some race conditions, perhaps related to storing these clients in state. You could either initialise everything at app start as opposed to button click, or move those `let client` outside of function into global scope so you access client from them and not state. But then again, this doesn't explain why builds work locally and not deployed, so it is more likely that something cache related is happening.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq I've figured out it and have found the exact reason! In my real project I have setting which supports es6 browserlist, so I've been investigating other reasons and have found out that the problem in _**modal initialization itself (new Web3Modal(...))**_. I moved modal initialization right before `openModal` call (see code in repo mentioned above) and modal became failing every time even on local env. You don't have this issue cause in your standalone example u initializes modal only once and uses it as an app constant, but I think _**it's not correct**_ since We (customers) don't want to make useless network requests on every page reload and the second point is that we can have other sign in providers than walletConnect and modal initialization becomes useless in this case.

**Comment by 0xAsimetriq** [comment]
That could be it! There is async ui import executed in modal's constructor, gonna see if I can find elegant solution to this without making modal initialisation async to avoid the race condition, good find 👍

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq Shall I create corresponding issue or will we continue the discussion in this thread?

**Comment by 0xAsimetriq** [comment]
Already working on this as part of next release 👍

**Comment by 0xAsimetriq** [comment]
@alexeygorbachevskiy could you please update and test @web3modal packages with this version `2.0.0-d8cdbe6`
Open method should now execute only after everything was loaded in.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq I'll try to test it later today.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq I've tested it on `2.0.0-d8cdbe6` and it seems to work fine. Can u answer my previous question about your solution for this?

**Comment by 0xAsimetriq** [comment]
In essence, when you configure the modal for the first time (new Web3Modal) there is an async import of ui library `await import('@web3modal/ui')` it usually takes few ms to resolve hence why there was no issues when initialising modal in the global scope, but you saw some race conditions as you called open method right after. It needs to be a conditional async import to ensure nothing breaks in server side rendered environments.

I didn't want to make modal's configuration async (it will complicate things for most common use case where it is configured early in app life cycle), so instead if open method is called before modal is initialised, it will be queued to execute after everything is ready.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq I think this is not a good solution as we need to rely on full initialization to take a couple of milliseconds or so, but what if not? It will also be bad if we need to show the loader before opening the modal, because again, when the loader stops spinning, we have no guarantee that the modal will be opened instantly.

**Comment by 0xAsimetriq** [comment]
I'll give this another look tomorrow to ensure there is an option for you to show loader reliably.

**Comment by alexeygorbachevskiy** [comment]
@0xAsimetriq Have you checked it?

**Comment by 0xAsimetriq** [comment]
@alexeygorbachevskiy Could you try out `2.0.0-e75d550`

You can now `await` modal's open method and show any necessary loaders with your approach.

**Comment by alexeygorbachevskiy** [comment]
Looks pretty good, but it would be better if we can `await` modal initialization, since when we `await openModal`, we can reliably hide spinner only after web3Modal was opened - it's not so good. But, anyway, thanks for the job, will wait for the next release.

**Comment by BestAssist** [comment]
@alexeygorbachevskiy Could you please open the standalone example again?
